openretina/models/linear_nonlinear.py

```diff
@@ -93,3 +98,188 @@ def forward(self, x: Float[torch.Tensor, "batch channels t h w"], data_key: str
+        else:
+            images = data_point.inputs
+        model_output = self.forward(images, session_id)
+        loss = self.loss.forward(model_output, data_point.targets) / sum(model_output.shape)
+        regularization = self.regularizer()
+        total_loss = loss + regularization
+        correlation = -self.correlation_loss.forward(model_output, data_point.targets)
+
+        self.log("val_loss", loss, logger=True, prog_bar=True)
+        self.log("val_regularization_loss", regularization, logger=True)
+        self.log("val_total_loss", total_loss, logger=True)
+        self.log("val_correlation", correlation, logger=True, prog_bar=True)
+
+        return loss
+
+    def configure_optimizers(self):
+        optimizer = torch.optim.AdamW(self.parameters(), lr=self.learning_rate)
+        lr_decay_factor = 0.3
+        patience = 5
+        tolerance = 0.0005
+        min_lr = self.learning_rate * (lr_decay_factor**3)
+        scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(
+            optimizer,
+            mode="max",
+            factor=lr_decay_factor,
+            patience=patience,
+            threshold=tolerance,
+            threshold_mode="abs",
+            min_lr=min_lr,
+        )
+        return {
+            "optimizer": optimizer,
+            "lr_scheduler": {
+                "scheduler": scheduler,
+                "monitor": "val_correlation",
+                "frequency": 1,
+            },
+        }
+
+
+    def laplace(self):
+        return self.smooth_weight_spat * self._smooth_reg_fn_spat(
+            self.space_conv.weight.squeeze(2)) + self.smooth_weight_temp * self._smooth_reg_fn_temp(
+            self.time_conv.weight.squeeze(-1,-2))
+
+    def weights_l1(self, average: bool = True):
+        """Returns l1 regularization across all weight dimensions
+
+        Args:
+            average (bool, optional): use mean of weights instead of sum. Defaults to True.
+        """
+        if average:
+            return self.space_conv.weight.abs().mean() + self.time_conv.weight.abs().mean()
+        else:
+            return self.space_conv.weight.abs().sum() + self.time_conv.weight.abs().sum()
+
+    def normalize_kernels(self):
+        """Normalizes the kernels to have unit norm."""
+        with torch.no_grad():
+            self.space_conv.weight.data /= self.space_conv.weight.data.norm(keepdim=True)
+            self.time_conv.weight.data /= self.time_conv.weight.data.norm(keepdim=True)
+
+    def forward(self, x: Float[torch.Tensor, "batch channels t h w"], data_key=None, **kwargs):
+        if self.normalize_weights:
+            self.normalize_kernels()
+
+        out = self.space_conv(x)
+        out = self.time_conv(out)
+        out = self.nonlinearity(out)
+        out = rearrange(out, "batch neurons t 1 1 -> batch t neurons")
+        return out
+
+    def regularizer(self):
+        return self.laplace() + self.sparse_weight * self.weights_l1()
```

Comment:
```suggestion
        return self.laplace() + self.sparse_weight * self.weights_l1()

```
normally you end each file in a new line, but I think running `make fix-all` should apply the codestyle automatically and also add this new line.